ocaml/xapi/cert_distrib.ml

```diff
@@ -364,15 +364,32 @@ let collect_pool_certs ~__context ~rpc ~session_id ~map ~from_hosts =
          map cert
      )
 
-let take_and_append n x xs =
-  (* take_and_append 3 10 [1;2;3;4] = [1;2;3;10] *)
-  let rec loop i acc = function
-    | x :: xs when i < n ->
-        loop (i + 1) (x :: acc) xs
-    | _ ->
-        x :: acc |> List.rev
+(* insert an element randomly *between* existing elements of a list.
+   We are enumerating the gaps between elements in a list:
+
+   List:     a b c d
+   Gaps:    0 1 2 3 4
+
+   A list of length n has gaps 0, 1, ...,n. We are selecting a gap
+   between 1 and n-1 to insert an element - so never at the beginning or
+   end if n >= 2. Otherwise we insert at the front.
+
+   This function is not tail recursive, so only apply to short lists *)
+
+let randomly_insert x ys =
+  let gap = Random.int (List.length ys) + 1 in
```

Comment:
To prevent the function from throwing an error when the length of the list is 0

Reply:
Because we never want to insert at 0 - the front, unless we can't avoid it. See the comment - we pick a gap from 1 .. n-1.

Reply:
Yes that's why I expected a `- 1`

Reply:
The call `Random.int 0` throws an `Invalid_argument`, this should be guarded against.

Reply:
Main point is that `gap` isn't actually the gap where we are inserting, which confused me. Because `gap` cannot be 0, but sometimes we do insert at 0; and `gap` can be = n, but we only insert at gaps < n.